File: android-exercises/11 - permissions/Permissions_java/app/src/main/java/com/bugrahankaramollaoglu/permissions_java/MainActivity.java
package com.bugrahankaramollaoglu.permissions_java;

import static android.app.ProgressDialog.show;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.ImageDecoder;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.View;
import android.widget.Toast;
import com.bugrahankaramollaoglu.permissions_java.databinding.ActivityMainBinding;
import com.google.android.material.snackbar.Snackbar;
import java.io.IOException;

public class MainActivity extends AppCompatActivity {

	private ActivityMainBinding binding;
	private ActivityResultLauncher<Intent> activityResultLauncher;
	private ActivityResultLauncher<String> permissionLauncher;
	Bitmap selectedBitmap;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		binding = ActivityMainBinding.inflate(getLayoutInflater());
		View view = binding.getRoot();
		setContentView(view);

		registerLauncher();
	}

	public void registerLauncher() {

		activityResultLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
				new ActivityResultCallback<ActivityResult>() {
					@Override
					public void onActivityResult(ActivityResult result) {
						if (result.getResultCode() == Activity.RESULT_OK) {
							Intent intentFromResult = result.getData();
							if (intentFromResult != null) {
								// UniformResourceIdentifier is a reference to the location of photo
								Uri imageData = intentFromResult.getData();
								try {
									selectedBitmap = MediaStore.Images.Media
											.getBitmap(MainActivity.this.getContentResolver(), imageData);
									binding.photoChosen.setImageBitmap(selectedBitmap);
								} catch (IOException e) {
									e.printStackTrace();
								}
							}
						}
					}
				});

		permissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestPermission(),
				new ActivityResultCallback<Boolean>() {
					@Override
					public void onActivityResult(Boolean result) {
						if (result) {
							Intent intentToGallery = new Intent(Intent.ACTION_PICK,
									MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
							activityResultLauncher.launch(intentToGallery);
						} else {
							Toast.makeText(MainActivity.this, "Permission needed!", Toast.LENGTH_LONG).show();
						}

					}
				});
	}

	public void selectImage(View view) {
		if (ContextCompat.checkSelfPermission(this,
				Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
			if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.READ_EXTERNAL_STORAGE)) {
				Snackbar.make(view, "permission needed to access gallery", Snackbar.LENGTH_INDEFINITE)
						.setAction("allow permission", new View.OnClickListener() {
							@Override
							public void onClick(View v) {
								permissionLauncher.launch(Manifest.permission.READ_EXTERNAL_STORAGE);
							}
						}).show();
			} else {
				permissionLauncher.launch(Manifest.permission.READ_EXTERNAL_STORAGE);
			}
		} else {
			Intent intentToGallery = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
			activityResultLauncher.launch(intentToGallery);
		}

	}
}

File: android-exercises/11 - permissions/Permissions_kotlin/app/src/main/java/com/example/permissions_kotlin/MainActivity.kt
package com.example.permissions_kotlin

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.ImageDecoder
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.MediaStore
import android.view.View
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.example.permissions_kotlin.databinding.ActivityMainBinding
import com.google.android.material.snackbar.Snackbar

class MainActivity : AppCompatActivity() {

    var selectedBitmap: Bitmap? = null
    private lateinit var binding: ActivityMainBinding
    private lateinit var activityResultLauncher: ActivityResultLauncher<Intent>
    private lateinit var permissionLauncher: ActivityResultLauncher<String>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        registerLauncher()
    }

    fun registerLauncher() {
        activityResultLauncher =
            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->

                if (result.resultCode == RESULT_OK) {
                    val intentFromResult = result.data
                    if (intentFromResult != null) {
                        val imageData = intentFromResult.data
                        try {

                            if (Build.VERSION.SDK_INT >= 28) {
                                val source = ImageDecoder.createSource(
                                    this@MainActivity.contentResolver, imageData!!
                                )
                                selectedBitmap = ImageDecoder.decodeBitmap(source)
                            } else {
                                selectedBitmap = MediaStore.Images.Media.getBitmap(
                                    this@MainActivity.contentResolver, imageData
                                )
                            }
                            binding.photoChosen.setImageBitmap(selectedBitmap)
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }
                }
            }

        permissionLauncher =
            registerForActivityResult(ActivityResultContracts.RequestPermission()) { result ->
                if (result) {
                    //permission granted
                    val intentToGallery =
                        Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
                    activityResultLauncher.launch(intentToGallery)
                } else {
                    //permission denied
                    Toast.makeText(this@MainActivity, "Permisson needed!", Toast.LENGTH_LONG).show()
                }
            }
    }

    fun selectImage(view: View) {
        if (ContextCompat.checkSelfPermission(
                this, Manifest.permission.READ_EXTERNAL_STORAGE
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(
                    this, Manifest.permission.READ_EXTERNAL_STORAGE
                )
            ) {
                Snackbar.make(view, "Permission needed for gallery", Snackbar.LENGTH_INDEFINITE)
                    .setAction("Give Permission", View.OnClickListener {
                        permissionLauncher.launch(Manifest.permission.READ_EXTERNAL_STORAGE)
                    }).show()
            } else {
                permissionLauncher.launch(Manifest.permission.READ_EXTERNAL_STORAGE)
            }

        } else {

            val intentToGallery =
                Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
            activityResultLauncher.launch(intentToGallery)
        }
    }
}
